Fatras/include/ActsFatras/Kernel/detail/Interactor.hpp

```diff
@@ -154,6 +182,118 @@ struct Interactor {
+        retval = true;
+      }
+      // the Interactor is in charge of keeping track of the material. since the
+      // accumulated material is stored in the particle it could (but should
+      // not) be modified by a physics process. to avoid issues, the material is
+      // updated only after process simulation has occured. this intentionally
+      // overwrites any material updates made by the process.
+      result.particle.setMaterialPassed(x0, l0);
+      return retval;
+    };
+
+    // material thickness measured in radiation/interaction lengths
+    const auto slabX0 = slab.thicknessInX0();
+    const auto slabL0 = slab.thicknessInL0();
+    // remaining radiation/interaction length to next point-like interaction
+    // NOTE for limit=inf this should result in dist=inf
+    const auto x0Dist = result.x0Limit - result.particle.pathInX0();
+    const auto l0Dist = result.l0Limit - result.particle.pathInL0();
+
+    // something point-like could happen within this material and we need to
+    // select which process would come first. x0/l0 measures the propagated path
+    // along different scales. to be able to check which one would happen first
+    // they need to be translated to a common scale.
+
+    // relative fraction within material where the interaction occurs.
+    //
+    // fraction < 0:
+    //   this is an error case where the point-like interaction should have
+    //   occured before reaching the material. not sure how this could happen,
+    //   but in such a case the point-like interaction happens immediately.
+    // 1 < fraction:
+    //   the next point-like interaction does not occur within the current
+    //   material. simulation is limited to the continuous processes.
+    //
+    // `clamp` ensures a valid range in all cases.
+    const float fracX0 = std::clamp(float(x0Dist / slabX0), 0.0f, 1.0f);
+    const float fracL0 = std::clamp(float(l0Dist / slabL0), 0.0f, 1.0f);
+    // fraction of the material where the first point-like interaction occurs
+    const float frac = std::min(fracX0, fracL0);
+
+    // do not run if there is zero material
+    if (0.0f < frac) {
+      // simulate continuous processes before the point-like interaction
+      if (runContinuousPartial(frac)) {
+        return;
+      }
+    }
+    // do not run if there is not point-like interaction
+    if (frac < 1.0f) {
+      // select which process to simulate
+      const size_t process =
+          (fracX0 < fracL0) ? result.x0Process : result.l0Process;
+      // simulate the selected point-like process
+      if (pointlike.run(*generator, process, result.particle,
+                        result.generatedParticles)) {
+        result.particleStatus = SimulationParticleStatus::eInteracted;
+        return;
+      }
+      // simulate continuous processes after the point-like interaction
+      if (runContinuousPartial(1.0 - frac)) {
+        return;
+      }
```

Comment:
Is the remaining material also considered for particles created by the interaction?

Reply:
No. I was first considering doing that, but realized that that would involve quite a bit of complex logic within the `Interactor`. You have to effectively run a mini-simulation there, since the daughter particles might spawn additional daughter particle that need to be handled. To have something working, I enforce a one-point-like-interaction-per-surface rule and assume the daughter particles are created at the end of the material.

That assumes that the effects of left-over material on generated particles are already included (to a certain extend) in the simulation models, e.g. in the parametrized nuclear interaction models that your are working.

Reply:
Might be good enough. Otherwise we just either pass the fraction to the propagation of the secondaries or store the secondaries here after applying the remaining material.